Tests/FBLPromisesTests/FBLPromise+RetryTests.m

```diff
@@ -0,0 +1,237 @@
+                       ? @42
+                       : [NSError errorWithDomain:FBLPromiseErrorDomain code:42 userInfo:nil];
+          }] then:^id(NSNumber *value) {
+    XCTAssertEqual(value, @42);
+    return nil;
+  }] catch:^(NSError *error) {
+    XCTFail(@"Promise should not be resolved with error.");
+  }];
+
+  // Assert.
+  XCTAssert(FBLWaitForPromisesWithTimeout(10.0));
+  XCTAssertEqual(count, 0);
+}
+
+- (void)testPromiseRetryRejectsBeforeRetryAttemptsAreExhaustedIfPredicateIsNotMet {
+  // Arrange.
+  NSInteger customAttempts = 3;
+  NSUInteger __block count = 1 + customAttempts;
+  NSError *retry42Error = [NSError errorWithDomain:FBLPromiseErrorDomain code:42 userInfo:nil];
+  NSError *retry13Error = [NSError errorWithDomain:FBLPromiseErrorDomain code:13 userInfo:nil];
+
+  // Act.
+  [[[FBLPromise attempts:customAttempts
+      delay:FBLPromiseRetryDefaultDelayInterval
+      condition:^BOOL(NSInteger remainingAttempts, NSError *error) {
+        XCTAssertEqual(count, remainingAttempts);
+        return error.code == retry42Error.code;
+      }
+      retry:^id {
+        count -= 1;
+        return count > 1 ? retry42Error : retry13Error;
+      }] then:^id(NSNumber *value) {
+    XCTFail(@"Promise should not be resolved with value.");
+    return nil;
+  }] catch:^(NSError *error) {
+    XCTAssertEqual(error, retry13Error);
+  }];
+
+  // Assert.
+  XCTAssert(FBLWaitForPromisesWithTimeout(10.0));
+  XCTAssertEqual(count, 1);
+}
+
+- (void)testPromiseRetryNoDeallocUntilResolved {
+  // Arrange.
+  FBLPromise *promise = [FBLPromise pendingPromise];
+  FBLPromise __weak *weakExtendedPromise1;
+  FBLPromise __weak *weakExtendedPromise2;
+  NSInteger customAttempts = 3;
+  NSUInteger __block count = 1 + customAttempts;
+  NSError *retry42Error = [NSError errorWithDomain:FBLPromiseErrorDomain code:42 userInfo:nil];
+
+  // Act.
+  @autoreleasepool {
+    XCTAssertNil(weakExtendedPromise1);
+    XCTAssertNil(weakExtendedPromise2);
+    weakExtendedPromise1 = [FBLPromise retry:^id {
+      return promise;
+    }];
+    weakExtendedPromise2 = [FBLPromise attempts:customAttempts
+        delay:FBLPromiseRetryDefaultDelayInterval
+        condition:^BOOL(NSInteger remainingAttempts, NSError *error) {
+          XCTAssertEqual(count, remainingAttempts);
+          return error.code == retry42Error.code;
+        }
+        retry:^id {
+          count -= 1;
+          return count > 1 ? retry42Error : promise;
+        }];
+    XCTAssertNotNil(weakExtendedPromise1);
+    XCTAssertNotNil(weakExtendedPromise2);
+  }
+
+  // Assert.
+  XCTAssertNotNil(weakExtendedPromise1);
+  XCTAssertNotNil(weakExtendedPromise2);
+
+  [promise fulfill:@42];
+  XCTAssert(FBLWaitForPromisesWithTimeout(10));
+  XCTAssertEqual(count, 1);
+  XCTAssertNil(weakExtendedPromise1);
+  XCTAssertNil(weakExtendedPromise2);
+}
+
+#pragma mark - Private
+
+/** Returns the time interval from `startDate` to current date rounded to one decimal place. */
```

Comment:
Rather than rounding time intervals and checking equality, isn't it better to assert that the time interval difference is < x ms (i.e. `0 < (delay - [date timeIntervalSinceDate:startDate]) < x`)?

Reply:
This actually brings up a good point that I wasn't considering. We can't really guarantee when the next dispatch async retry attempt will be made. That's up to current state of GCD, which is handling the dispatch management for us. We can really only guarantee that the attempt will be made after the specified delay interval. Updated the tests to reflect that.